File: backend/src/Schedulas.Application/Features/Calendar/CalendarDtos.cs
using Schedulas.Domain.Enums;

namespace Schedulas.Application.Features.Calendar;

public sealed record CalendarEntryDto
{
    public Guid ActivityId { get; init; }
    public string Title { get; init; } = default!;
    public string? Description { get; init; }
    public DateOnly Date { get; init; }
    public TimeOnly? StartTime { get; init; }
    public TimeOnly? EndTime { get; init; }
    public TimeSpan? Duration { get; init; }
    public ActivityStatus Status { get; init; }
    public int Priority { get; init; }
    public ActivityType ActivityType { get; init; }
    
    public Guid ClassId { get; init; }
    public string ClassName { get; init; } = default!;
    public string CourseName { get; init; } = default!;
    
    // Metadata derived
    public string? TeacherId { get; init; }
    public string? TeacherName { get; init; }
    public string? RoomId { get; init; }
    public string? RoomName { get; init; }
    
    // Derived UI Color
    public string ColorHex { get; init; } = default!;
}

public sealed record DailyCalendarDto
{
    public DateOnly Date { get; init; }
    public IReadOnlyList<CalendarEntryDto> Entries { get; init; } = [];
}

public sealed record WeeklyCalendarDto
{
    public DateOnly StartDate { get; init; }
    public DateOnly EndDate { get; init; }
    public IReadOnlyList<DailyCalendarDto> Days { get; init; } = [];
}

public sealed record MonthlyCalendarDto
{
    public int Year { get; init; }
    public int Month { get; init; }
    public IReadOnlyList<WeeklyCalendarDto> Weeks { get; init; } = [];
}

File: flutter_app/lib/features/calendar/domain/models/calendar_dtos.dart
import '../../../activities/domain/models/activity_enums.dart';

class CalendarEntryDto {
  final String activityId;
  final String title;
  final String? description;
  final String date;
  final String? startTime;
  final String? endTime;
  final String? duration;
  final ActivityStatus status;
  final int priority;
  final ActivityType activityType;

  final String classId;
  final String className;
  final String courseName;

  final String? teacherId;
  final String? teacherName;
  final String? roomId;
  final String? roomName;

  final String colorHex;

  CalendarEntryDto({
    required this.activityId,
    required this.title,
    this.description,
    required this.date,
    this.startTime,
    this.endTime,
    this.duration,
    required this.status,
    required this.priority,
    required this.activityType,
    required this.classId,
    required this.className,
    required this.courseName,
    this.teacherId,
    this.teacherName,
    this.roomId,
    this.roomName,
    required this.colorHex,
  });

  factory CalendarEntryDto.fromJson(Map<String, dynamic> json) {
    return CalendarEntryDto(
      activityId: json['activityId'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      date: json['date'] as String,
      startTime: json['startTime'] as String?,
      endTime: json['endTime'] as String?,
      duration: json['duration'] as String?,
      status: ActivityStatusExtension.fromValue(json['status'] as int),
      priority: json['priority'] as int,
      activityType:
          ActivityTypeExtension.fromValue(json['activityType'] as int),
      classId: json['classId'] as String,
      className: json['className'] as String,
      courseName: json['courseName'] as String,
      teacherId: json['teacherId'] as String?,
      teacherName: json['teacherName'] as String?,
      roomId: json['roomId'] as String?,
      roomName: json['roomName'] as String?,
      colorHex: json['colorHex'] as String,
    );
  }
}

class DailyCalendarDto {
  final String date;
  final List<CalendarEntryDto> entries;

  DailyCalendarDto({
    required this.date,
    required this.entries,
  });

  factory DailyCalendarDto.fromJson(Map<String, dynamic> json) {
    return DailyCalendarDto(
      date: json['date'] as String,
      entries: (json['entries'] as List<dynamic>?)
              ?.map((e) => CalendarEntryDto.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class WeeklyCalendarDto {
  final String startDate;
  final String endDate;
  final List<DailyCalendarDto> days;

  WeeklyCalendarDto({
    required this.startDate,
    required this.endDate,
    required this.days,
  });

  factory WeeklyCalendarDto.fromJson(Map<String, dynamic> json) {
    return WeeklyCalendarDto(
      startDate: json['startDate'] as String,
      endDate: json['endDate'] as String,
      days: (json['days'] as List<dynamic>?)
              ?.map((e) => DailyCalendarDto.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class MonthlyCalendarDto {
  final int year;
  final int month;
  final List<WeeklyCalendarDto> weeks;

  MonthlyCalendarDto({
    required this.year,
    required this.month,
    required this.weeks,
  });

  factory MonthlyCalendarDto.fromJson(Map<String, dynamic> json) {
    return MonthlyCalendarDto(
      year: json['year'] as int,
      month: json['month'] as int,
      weeks: (json['weeks'] as List<dynamic>?)
              ?.map(
                  (e) => WeeklyCalendarDto.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}
